core/src/main/java/org/lflang/generator/docker/DockerGenerator.java

```diff
@@ -24,31 +32,174 @@ public DockerGenerator(LFGeneratorContext context) {
+                .flatMap(java.util.Collection::stream)
+                .collect(Collectors.toList()),
+            " \\\n\t&& ");
+  }
+
+  /** Return the ENTRYPOINT command. */
+  protected String generateEntryPoint() {
+    return "ENTRYPOINT ["
+        + getEntryPointCommands().stream()
+            .map(cmd -> "\"" + cmd + "\"")
+            .collect(Collectors.joining(","))
+        + "]";
+  }
+
+  /** Return a COPY command to copy the executable from the builder to the runner. */
+  protected String generateCopyOfExecutable() {
+    var lfModuleName = context.getFileConfig().name;
+    return "COPY --from=builder /lingua-franca/%s/bin/%s ./bin/%s"
+        .formatted(lfModuleName, lfModuleName, lfModuleName);
   }
 
-  public String userRunCommand() {
-    var runCmd = context.getTargetConfig().get(DockerProperty.INSTANCE).run();
-    if (runCmd != null && !runCmd.isEmpty()) {
-      return "RUN " + runCmd;
+  /** Return a COPY command to copy the scripts from the builder to the runner. */
+  protected String generateCopyOfScript() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preRunScript();
+    if (!script.isEmpty()) {
+      return "COPY --from=builder /lingua-franca/%s/src-gen/%s ./scripts/%s"
+          .formatted(context.getFileConfig().name, script, script);
+    }
+    return "# (No pre-run script provided.)";
+  }
+
+  /**
+   * Return a list of strings used to construct and entrypoint. If this is done for a federate, then
+   * also include additional parameters to pass in the federate ID.
+   */
+  protected List<String> entryPoint() {
+    if (context instanceof SubContext) {
+      return Stream.concat(defaultEntryPoint().stream(), List.of("-i", "1").stream()).toList();
     } else {
-      return "";
+      return defaultEntryPoint();
+    }
+  }
+
+  /**
+   * Return a list of commands to be used to construct an ENTRYPOINT, taking into account the
+   * existence of a possible pre-run script.
+   */
+  protected final List<String> getEntryPointCommands() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preRunScript();
+    if (!script.isEmpty()) {
+      return List.of(
+          DockerOptions.DEFAULT_SHELL,
+          "-c",
+          "source scripts/"
+              + script
+              + " && "
+              + entryPoint().stream().collect(Collectors.joining(" ")));
     }
+    return entryPoint();
+  }
+
+  /** The default list of commands to construct an ENTRYPOINT out of. Different for each target. */
+  public abstract List<String> defaultEntryPoint();
+
+  /** Return the default base image. */
+  public abstract String defaultImage();
+
+  /** Return the base image to be used during the building stage. */
+  protected String builderBase() {
+    return baseImage(context.getTargetConfig().get(DockerProperty.INSTANCE).builderBase());
+  }
+
+  /** Return the base image to be used during the running stage. */
+  protected String runnerBase() {
```

Comment:
Would it be helpful for this to default to builderBase? I assume that cross-compiling is a sufficiently uncommon use case that having the two be the same by default would be the least surprising behavior for users.

Reply:
Yes, I thought about this too. Should be an easy fix, so I'll do that.